falcon/testing/smmock.py

```diff
@@ -0,0 +1,139 @@
+    the steps to be called individually in order to test the middleware
+    outside of a normal WSGI call where the `process_response` may clean
+    up values setup earlier by `process_request` or `process_resource`.
+
+    MiddlewareMock applies the same rules on the middleware that the
+    `falcon.API` does. That is, if `middleware` is a list of objects, then
+    `process_response` runs the middleware in reverse order of either
+    `process_request` or `process_resource` thereby simulating the order in
+    which the middleware would run in a normal operation.
+
+    Attributes:
+        middleware (object or list, required): One or more objects
+            (instantiated classes) that implement the Falcon Middleware
+            specification. See `falcon.API` for details.
+        env (dict): environment dictionary used to create the falcon.Request
+            instance used for tests.
+        path (string): URI path used for the `falcon.Request`
+        protocol (str): HTTP protocol version, default 'HTTP/1.0'
+        request_options (dict, optional): Options to pass to `falcon.Request`
+        request (falcon.Request): instance of `falcon.Request` to use. If it
+            is None or not an instance of `falcon.Request` then it will be
+            created.
+        resource (object): object instance that provides the resource that the
+            routing would have normally provided. This is passed directly
+            to the middleware for `process_resource` and `process_response`.
+            It does not apply to `process_request`.
+        response (falcon.Response): instance of `falcon.Response` to use. If it
+            is None or not an instance of `falcon.Response` then it will be
+            created.
+
+    """
+
+    def __init__(self):
+        self.middleware = None
+        self.env = None
+        self.path = None
+        self.protocol = 'HTTP/1.0'
+        self.request_options = None
+        self.request = None
+        self.resource = None
+        self.response = None
+
+    def _build(self, **kwargs):
+        """Create the request and response information"""
+        if self.env is None:
+            self.env = create_environ(
+                path=self.path,
+                protocol=self.protocol,
+                **kwargs)
+
+        # validate that self.request is an instance of falcon.Request
+        # or reset it
+        if self.request is not None:
+            if not isinstance(self.request, Request):
+                self.request = None
+        if self.request is None:
+            self.request = Request(self.env, self.request_options)
+
+        # validate that self.response is an instance of falcon.Response
+        # or reset it
+        if self.response is not None:
+            if not isinstance(self.response, Response):
+                self.response = None
+        if self.response is None:
+            self.response = Response()
+
+    def simulate_process_request(self, **kwargs):
```

Comment:
How do isolate a middleware test to verify it is a bug in just the middleware?

Testing the middleware using the full falcon.API seems at least to me to be an integration test instead of a unit test which is the purpose of these additional functions.